source/extensions/common/wasm/plugin.cc

```diff
@@ -20,6 +16,45 @@ WasmConfig::WasmConfig(const envoy::extensions::wasm::v3::PluginConfig& config)
     // TODO(rapilado): Set the SanitizationConfig fields once sanitization is implemented.
     allowed_capabilities_[capability.first] = proxy_wasm::SanitizationConfig();
   }
+
+  if (config_.vm_config().has_environment_variables()) {
+    const auto& envs = config_.vm_config().environment_variables();
+
+    // We reject NullVm with key_values configuration
+    // since it directly accesses Envoy's env vars and we should not modify Envoy's env vars here.
+    // TODO(mathetake): Once proxy_get_map_values(type::EnvironmentVariables, ..) call is supported,
+    // then remove this restriction.
+    if (config.vm_config().runtime() == WasmRuntimeNames::get().Null &&
+        !envs.key_values().empty()) {
+      throw EnvoyException("envoy.extensions.wasm.v3.VmConfig.EnvironmentVariables.key_values must "
+                           "not be set for NullVm.");
+    }
+
+    // Check key duplication.
+    absl::flat_hash_set<std::string> keys;
+    for (const auto& env : envs.key_values()) {
+      keys.insert(env.first);
+    }
+    for (const auto& key : envs.host_env_keys()) {
+      if (!keys.insert(key).second) {
```

Comment:
While this might work in this particular case, because you're inserting only keys first, this is not the correct way to check for presence of the existing key.

I think @htuch meant that you can replace this whole block something like this (untested):
```
// Copy key-value pairs.
envs_ = envs.key_values();

// Copy variables from the host environment.
for (const auto& key : envs.host_env_keys()) {
  if (auto value = std::getenv(key.data())) {
    const auto [it, success] = envs_.insert({key, value});
    if (!success) {
      throw EnvoyException(...);
    }
  }
}
```

Reply:
We need to check duplication in the config regardless of existence of host env vars (following your comment https://github.com/envoyproxy/envoy/pull/15136#discussion_r586845896). And this is why I have a separate block with hash_set, otherwise we might end up inserting env vars with empty string value.

Aslo key_values() is Map of proto so there's no conversion to std::map.

I would appreciate it if there is better workaround.

Reply:
Oops, right! Clearly, you have a better memory than myself.